dags/challenge_dag_factory.py

```diff
@@ -0,0 +1,209 @@
+import io
+import requests
+import uuid
+from datetime import datetime
+import yaml
+import pandas as pd
+
+from airflow.decorators import dag, task
+from airflow.models import Param
+from airflow.providers.amazon.aws.hooks.s3 import S3Hook
+
+from orca.services.nextflowtower import NextflowTowerHook
+from orca.services.nextflowtower.models import LaunchInfo
+from orca.services.synapse import SynapseHook
+
+# Define the path to your challenge configuration file.
+CONFIG_URL = "https://raw.githubusercontent.com/Sage-Bionetworks-Workflows/orca-recipes/main/dags/challenge_configs.yaml"
+
+
+def load_challenge_configs(url=CONFIG_URL):
+    """Load challenge configurations from a raw GitHub URL."""
+
+    response = requests.get(url)
+
+    if not response.ok:
+            raise RuntimeError(
+                f"Failed to fetch challenge configs... "
+                f"Status code: {response.status_code}, reason: {response.reason}"
+            )
+
+    return yaml.safe_load(response.text)
+
+def resolve_dag_config(challenge_name: str, dag_params: dict, config: dict) -> dict:
+    """
+    Return the DAG configuration for a challenge.
+
+    If the challenge configuration provides a custom `dag_config`, use it
+    (ensuring any ISO-format dates are converted and that task parameters are injected).
+    Otherwise, return the default DAG configuration.
+
+    Arguments:
+        challenge_name: The name of the challenge.
+        dag_params: A dictionary of DAG parameters.
+        config: The challenge configuration.
+
+    Returns:
+        dict: The resolved DAG configuration.
+    """
+    
+    # Start with default configuration
+    dag_config = {
+        "schedule_interval": "*/1 * * * *",
+        "start_date": datetime(2024, 4, 9),
+        "catchup": False,
+        "default_args": {"retries": 2},
+        "tags": ["nextflow_tower"],
+        "params": dag_params,
+    }
+
+    # Update with any custom configuration if provided
+    if config.get('dag_config'):
+        dag_config.update(config['dag_config'])
+        
+        # Ensure start_date is a datetime object if provided
+        if 'start_date' in dag_config and isinstance(dag_config['start_date'], str):
+            dag_config['start_date'] = datetime.fromisoformat(dag_config['start_date'])
+            
+        # Ensure challenge name is in tags
+        if 'tags' in dag_config:
+            if challenge_name not in dag_config['tags']:
+                dag_config['tags'].append(challenge_name)
```

Comment:
I am curious why is it important for the challenge name to be in "tag"? is it for filtering purpose?

Reply:
Yep! For filtering purposes in the UI. We expect a single challenge to have multiple DAG configuration profiles. Either because it has multiple rounds, or multiple task types, or both. So filtering for a given challenge in the UI so I can easily pause/unpause between rounds and tasks will make things easier for me.